### PR TITLE
Fix ad totals should update on month roll over without fetching latest balances from server

### PR DESCRIPTION
**Pull request by tmancey**
Fixes https://github.com/brave/brave-browser/issues/4994

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

https://docs.google.com/document/d/1qGJKVXUtdx4cPFDW5kU2cXbfUmwrC_dh0dE1gUFIVKA/edit#

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc [hunk 1]
```diff
@@ -177,6 +250,28 @@ bool Payments::GetBalanceFromDictionary(
...
+  DCHECK(month);
+
+  auto* value = dictionary->FindKey("month");
+  if (!value || !value->is_string()) {
+    return false;
+  }
+
+  auto month_value = value->GetString();
+
+  if (!re2::RE2::FullMatch(month_value, "[0-9]{4}-[0-9]{2}")) {
```
Comment (emerick) [inline vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc]: Nit: A comment here might be helpful; took me a second to realize that this regex was looking for a year and a month (was expecting just a month, based on the variable/function names).

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc [hunk 2]
```diff
@@ -199,30 +294,48 @@ bool Payments::GetTransactionCountFromDictionary(
...
+}
+
+std::string Payments::GetPreviousTransactionMonth(
+    const base::Time& time) const {
+  base::Time::Exploded time_exploded;
+  time.LocalExplode(&time_exploded);
+
+  time_exploded.month--;
+  if (time_exploded.month < 1) {
+    time_exploded.month += 12;
```
Comment (emerick) [inline vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc]: I think `time_exploded.month = 12` makes the intent a little clearer here, even though it's the overall same effect.

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc (hunk 1 above)
Comment (tmancey) [reply vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc]: 👍 fixed

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc (hunk 2 above)
Comment (tmancey) [reply vendor/bat-native-confirmations/src/bat/confirmations/internal/payments.cc]: 👍 fixed

**Pull request merged**